### PR TITLE
refactor(pagination): First try on the BEM naming of a widget

### PR DESCRIPTION
**Pull request by pixelastic**
I added BEM naming to the pagination widget. The generated markup
looks like:

```html
<ul class="ais-pagination">
  <li class="ais-pagination--item ais-pagination--item-first ais-pagination--item__disabled"><a href="#">First</a></li>
  <li class="ais-pagination--item ais-pagination--item-previous ais-pagination--item__disabled"><a href="#">Previous</a></li>
  <li class="ais-pagination--item ais-pagination--item-page ais-pagination--item-page__active"><a href="#">1</a></li>
  <li class="ais-pagination--item ais-pagination--item-page"><a href="#">2</a></li>
  <li class="ais-pagination--item ais-pagination--item-page"><a href="#">3</a></li>
  <li class="ais-pagination--item ais-pagination--item-page"><a href="#">4</a></li>
  <li class="ais-pagination--item ais-pagination--item-page"><a href="#">5</a></li>
  <li class="ais-pagination--item ais-pagination--item-next"><a href="#">Last</a></li>
  <li class="ais-pagination--item ais-pagination--item-last"><a href="#">Last</a></li>
</ul>
```

Everytime I add a `BEM` class, one can add its own set of custom
classes through the `cssClasses` object config in the widget.
A default `pagination.css` file is provided but currently only
includes the `visibility:hidden` styling of disabled elements. The
demo styling is thus done through bootstrap classes.

I quite happy with the end result. I also simplified the existing code
in the process and moved all default attributes in the widget (instead
of the components).

I'm submitting the PR to get feedback and then I'll update the docs
and maybe add some default styling not from Bootstrap.

### CODE REVIEW

**Comment by vvo** [comment]
I like it, we then need to make sure to document in the readme what are the available classes (just copy pasting the default css should be enough).

File: widgets/pagination.js [hunk 1]
```diff
@@ -1,50 +1,85 @@
...
+    labels = {},
+    maxPages = 20,
+    padding = 3,
+    showFirstLast = true,
     hideWhenNoResults = true
   }) {
   var containerNode = utils.getContainerNode(container);
 
+  if (!container) {
+    throw new Error('Usage: pagination({container[, cssClasses, labels.{prev,next,first,last}, maxPages, showFirstLast, hideWhenNoResults]})');
```
Comment (vvo) [inline widgets/pagination.js]: cssClasses and labels need adjustement?

**Comment by vvo** [comment]
This should provide a very good extensibility feature for people needing so. :100:

File: components/Pagination/Pagination.js [hunk 2]
```diff
@@ -14,88 +14,97 @@ class Pagination extends React.Component {
...
+    if (isDisabled) {
+      className = cx(bem('item', 'disabled'), this.props.cssClasses.disabled, className);
+    }
+    if (isActive) {
+      className = cx(bem('item-page', 'active'), this.props.cssClasses.active, className);
     }
 
+
     return (
-      <PaginationLink
```
Comment (bobylito) [inline components/Pagination/Pagination.js]: Why did you remove the PaginationLink component?

Comment (vvo) [inline components/Pagination/Pagination.js]: Yup I think we could still benefit from having a separate `<PaginationLink` component here.

File: widgets/pagination.js (hunk 1 above)
Comment (pixelastic) [inline widgets/pagination.js]: I'd really like a way to do this check automatically...

File: components/Pagination/Pagination.js (hunk 2 above)
Comment (pixelastic) [inline components/Pagination/Pagination.js]: Because it was only passing the props to a `li > a`, all logic was already handled in the `Pagination` component. I thought this was one layer too many. I don't see the value of it, could you explain?

Comment (bobylito) [inline components/Pagination/Pagination.js]: The building we're using here is mostly the component. So if you have a something that is big enough for its own function it should go in a component.

Eventually when we move to 0.14 we can start creating function based components.

**Comment by vvo** [comment]
Are we merging this and waiting for 0.14 to do a refactor on pageLink being a simple function component?

**Comment by bobylito** [comment]
IMHO it should have stayed a component in the first place. I would revert that into a standalone component and later migrate to a simple function component.

**Comment by vvo** [comment]
> IMHO it should have stayed a component in the first place. I would revert that into a standalone component and later migrate to a simple function component.

WFM @pixelastic care about re-putting it as a component? Indeed in the end this is not part of the feature you are adding (BEM), so let's not try to spend time on this subject (right now) while the added BEM functionnality is good :)

**Comment by pixelastic** [comment]
Yep, I'll push that soon.

**Comment by bobylito** [comment]
> so let's not try to spend time on this subject (right now) while the added BEM functionnality is good :)

Yep true @vvo, this feature is really awesome @pixelastic  :dango: :dancers:

**Comment by pixelastic** [comment]
I've re-added the PaginationLink component, ready to be merged.

**Comment by bobylito** [comment]
Lgtm :)
